receiver/snmpreceiver/scraper.go

```diff
@@ -39,13 +64,345 @@ func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCrea
+	s.client, err = newClient(s.cfg, s.logger)
+
+	return err
 }
 
 // scrape collects and creates OTEL metrics from a SNMP environment
 func (s *snmpScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
-	md := pmetric.NewMetrics()
+	if err := s.client.Connect(); err != nil {
+		return pmetric.NewMetrics(), fmt.Errorf("problem connecting to SNMP host: %w", err)
+	}
+	defer s.client.Close()
+
+	// Create the metrics helper which will help manage a lot of the otel metric and resource functionality
+	metricHelper := newOTELMetricHelper(s.settings)
+
+	configHelper := newConfigHelper(s.cfg)
+
+	var scraperErrors scrapererror.ScrapeErrors
+	// Try to scrape scalar OID based metrics
+	s.scrapeScalarMetrics(metricHelper, configHelper, &scraperErrors)
+
+	// Try to scrape column OID based metrics
+	s.scrapeIndexedMetrics(metricHelper, configHelper, &scraperErrors)
+
+	return metricHelper.metrics, scraperErrors.Combine()
+}
+
+// scrapeScalarMetrics retrieves all SNMP data from scalar OIDs and turns the returned scalar data
+// into metrics with optional enum attributes
+func (s *snmpScraper) scrapeScalarMetrics(
+	metricHelper *otelMetricHelper,
+	configHelper *configHelper,
+	scraperErrors *scrapererror.ScrapeErrors,
+) {
+	metricScalarOIDs := configHelper.getMetricScalarOIDs()
+
+	// If no scalar metric configs, nothing else to do
+	if len(metricScalarOIDs) == 0 {
+		return
+	}
+
+	// Retrieve all SNMP data from scalar metric OIDs
+	scalarData := s.client.GetScalarData(metricScalarOIDs, scraperErrors)
+
+	// Create general resource if we're about to make scalar metrics
+	resource := metricHelper.getResource(generalResourceKey)
+	if resource == nil {
+		metricHelper.createResource(generalResourceKey, map[string]string{})
+	}
+
+	// For each piece of SNMP data, attempt to create the necessary OTEL structures (resources/metrics/datapoints)
+	for _, data := range scalarData {
+		if err := s.scalarDataToMetric(data, metricHelper, configHelper); err != nil {
+			scraperErrors.AddPartial(1, fmt.Errorf(errMsgScalarOIDProcessing, data.oid, err))
+		}
+	}
+}
+
+// scrapeIndexedMetrics retrieves all SNMP data from column OIDs and turns the returned indexed data
+// into metrics with optional attribute and/or resource attributes
+func (s *snmpScraper) scrapeIndexedMetrics(
+	metricHelper *otelMetricHelper,
+	configHelper *configHelper,
+	scraperErrors *scrapererror.ScrapeErrors,
+) {
+	// Retrieve column OID SNMP indexed data for attributes
+	indexedAttributeValues := s.scrapeIndexedAttributes(configHelper.getAttributeColumnOIDs(), scraperErrors)
+
+	// Retrieve column OID SNMP indexed data for resource attributes
+	indexedResourceAttributeValues := s.scrapeIndexedAttributes(configHelper.getResourceAttributeColumnOIDs(), scraperErrors)
+
+	metricColumnOIDs := configHelper.getMetricColumnOIDs()
+
+	// If no column metric configs, nothing else to do
+	if len(metricColumnOIDs) == 0 {
+		return
+	}
```

Comment:
Let's do this before scraping attribute values, since we'll only use them if this condition passes.

Reply:
@djaglowski i feel like this was intentionally placed at one point....but i can't think of why. at least it doesn't seem to serve any purpose anymore to have this come later. i'll make this change